File: scripts/generate-qr.ts
#!/usr/bin/env -S node --loader tsx
import { parse } from 'csv-parse/sync';
import fs from 'fs';
import path from 'path';
import QRCode from 'qrcode';

type Args = {
  input: string;
  outDir: string;
  payload: 'url' | 'id' | 'json';
  baseUrl?: string; // required when payload=url
};

function parseArgs(): Args {
  const argv = process.argv.slice(2);
  let input = '';
  let outDir = './qr-out';
  let payload: Args['payload'] = 'url';
  let baseUrl: string | undefined = process.env.QR_BASE_URL;

  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--in' || a === '--input') input = argv[++i];
    else if (a === '--out' || a === '--outDir') outDir = argv[++i];
    else if (a === '--payload') payload = argv[++i] as Args['payload'];
    else if (a === '--base' || a === '--baseUrl') baseUrl = argv[++i];
  }

  if (!input) {
    console.error('Usage: tsx scripts/generate-qr.ts --in <csv> [--out ./qr-out] [--payload url|id|json] [--base https://your-domain/qr]');
    process.exit(1);
  }
  if (payload === 'url' && !baseUrl) {
    console.error('When --payload url is used, provide --base https://your-domain/qr or set QR_BASE_URL env var');
    process.exit(1);
  }
  return { input, outDir, payload, baseUrl };
}

function ensureDir(dir: string) {
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
}

async function main() {
  const { input, outDir, payload, baseUrl } = parseArgs();
  const csv = fs.readFileSync(input, 'utf8');
  const rows: any[] = parse(csv, { columns: true, skip_empty_lines: true });

  ensureDir(outDir);

  // The dataset has continuation rows (empty fitting_id). Filter for primary rows with qr_code__qr_id
  const primaries = rows.filter(r => r['fitting_id'] && r['qr_code__qr_id']);

  let count = 0;
  for (const row of primaries) {
    const qrId: string = row['qr_code__qr_id'];
    let text = '';
    if (payload === 'url') text = `${baseUrl!.replace(/\/$/, '')}?id=${encodeURIComponent(qrId)}`;
    else if (payload === 'id') text = qrId;
    else text = JSON.stringify({ qr_id: qrId, fitting_id: row['fitting_id'] });

    const filename = path.join(outDir, `${qrId}.png`);
    await QRCode.toFile(filename, text, {
      type: 'png',
      errorCorrectionLevel: 'M',
      margin: 1,
      width: 1024,
      color: { dark: '#000000', light: '#FFFFFFFF' },
    });
    count++;
  }
  console.log(`Generated ${count} QR PNGs in ${outDir}`);
}

main().catch((e) => { console.error(e); process.exit(1); });

File: scripts/generate_qr.py
#!/usr/bin/env python3
import argparse
import csv
import os
from pathlib import Path

import qrcode


def build_payload(qr_id: str, fitting_id: str, mode: str, base_url: str | None) -> str:
    if mode == 'url':
        if not base_url:
            raise SystemExit('When mode=url, provide --base https://your-domain/qr')
        return f"{base_url}?id={qr_id}"
    if mode == 'json':
        import json
        return json.dumps({'qr_id': qr_id, 'fitting_id': fitting_id})
    return qr_id


def main():
    ap = argparse.ArgumentParser(description='Generate QR PNGs from CSV with qr_code__qr_id')
    ap.add_argument('--in', dest='input', required=True, help='Path to CSV (Dataset/track_fittings_dataset.csv)')
    ap.add_argument('--out', dest='out', default='qr-out', help='Output directory')
    ap.add_argument('--mode', dest='mode', default='url', choices=['url', 'id', 'json'])
    ap.add_argument('--base', dest='base', default=os.getenv('QR_BASE_URL'))
    args = ap.parse_args()

    out_dir = Path(args.out)
    out_dir.mkdir(parents=True, exist_ok=True)

    count = 0
    with open(args.input, newline='', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            if not row.get('fitting_id') or not row.get('qr_code__qr_id'):
                continue
            qr_id = row['qr_code__qr_id']
            payload = build_payload(qr_id, row['fitting_id'], args.mode, args.base)
            img = qrcode.make(payload)
            img.save(out_dir / f"{qr_id}.png")
            count += 1

    print(f"Generated {count} QR PNGs in {out_dir}")


if __name__ == '__main__':
    main()


